lib/src/realm_class.dart

```diff
@@ -106,15 +106,58 @@ class Realm implements Finalizable {
   /// and will not update when writes are made to the database.
   late final bool isFrozen;
 
+  /// Returns true if the [Realm] is opened for the first time and the realm file is created.
+  late final bool _openedFirstTime;
+
   /// Opens a `Realm` using a [Configuration] object.
   Realm(Configuration config) : this._(config);
 
-  Realm._(this.config, [RealmHandle? handle, this._isInMigration = false]) : _handle = handle ?? _openRealm(config) {
+  Realm._(this.config, [RealmHandle? handle, this._isInMigration = false])
+      : _openedFirstTime = !File(config.path).existsSync(),
+        _handle = handle ?? _openRealmSync(config) {
     _populateMetadata();
     isFrozen = realmCore.isFrozen(this);
   }
 
-  static RealmHandle _openRealm(Configuration config) {
+  /// A method for asynchronously opening a [Realm].
+  ///
+  /// If the configuration is [FlexibleSyncConfiguration], the realm will be downloaded and fully
+  /// synchronized with the server prior to the completion of the returned [Future].
+  /// This method could be called also for opening a local [Realm].
+  ///
+  /// * `config`- a configuration object that describes the realm.
+  /// * `cancellationToken` - an optional [CancellationToken] used to cancel the operation.
+  /// * `onProgressCallback` - a callback for receiving download progress notifications for synced [Realm]s.
+  ///
+  /// Returns [Future<Realm>] that completes with the `realm` once the remote realm is fully synchronized or with an `error` if operation is canceled.
+  /// When the configuration is [LocalConfiguration] this completes right after the local realm is opened or operation is canceled.
+  static Future<Realm> open(Configuration config, {CancellationToken? cancellationToken, ProgressCallback? onProgressCallback}) async {
+    Realm realm = await CancellableFuture.fromFutureFunction<Realm>(() => _open(config, cancellationToken, onProgressCallback), cancellationToken);
+    return realm;
+  }
+
+  static Future<Realm> _open(Configuration config, CancellationToken? cancellationToken, ProgressCallback? onProgressCallback) async {
+    Realm realm = Realm(config);
+    cancellationToken?.onBeforeCancel(() async {
+      realm.close();
+    });
+
+    if (config is FlexibleSyncConfiguration) {
+      final session = realm.syncSession;
+      if (onProgressCallback != null) {
+        await session
+            .getProgressStream(ProgressDirection.download, ProgressMode.forCurrentlyOutstandingWork)
+            .forEach((s) => onProgressCallback.call(s.transferredBytes, s.transferableBytes));
+      }
+      await session.waitForDownload();
+      if (!realm._openedFirstTime) {
+        await session.waitForUpload();
+      }
```

Comment:
Why are we doing this?

Reply:
Can you check 2 comments above. I was wondering whether we need to wait for upload in case of having existing realm that has been modified offline and then opened async?

Reply:
Ah missed that. replied in the correct thread.